docs/csharp/language-reference/compiler-messages/cs8410.md

```diff
@@ -0,0 +1,34 @@
+---
+title: Compiler Error CS8410
+ms.date: 07/11/2020
+f1_keywords:
+  - "CS8410"
+helpviewer_keywords:
+  - "CS8410"
+author: Youssef1313
+---
+# Compiler Error CS8410
+
+'type': type used in an asynchronous using statement must be implicitly convertible to 'System.IAsyncDisposable' or implement a suitable 'DisposeAsync' method.
+
+The expression inside an `await using` statement must have a `DisposeAsync` method.
+
+## To correct this error
+
+Use `using` instead of `await using`.
+
+## Example
+
+<!-- TODO: The current example is a little bit poor. -->
```

Comment:
I think a really good example would be using `StreamReader`, because `StreamWriter` does have a `DisposeAsync` method, but `StreamReader` does not.

Reply:
I tried StreamReader but it generated CS8417 instead of CS8410 (Both errors are similar). Not sure when the compiler generates which one.

Reply:
Is this supposed to be an example of code that doesn't compile? If so, you'd show an example of something that is errant.

[Here is a SharpLab.io] (https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQBgAIUMA6ASQHkBuXA4jAVlpzowDZi0GB2Q3AN65CIhpxQAOYpwCyEAJYA7ABQBKfjlGEhmraJQBOYkWUA3CAkII4EACaJCAXkKK4Ad0IBlGNYhgASjb2CMoARKGq6gD0UYQAIgD2cFCKAOQwhAAWEKZwhPIwwnoiMYbGhGYWVkEOzq4e3r5gAOoIBYhhEdGx2VDx8lAADglQcACCUACeigDGRcU6xSIAvvOEqzjLQA) showing that the `StreamWriter` can be used with `await using` but the `StreamReader` cannot.

Reply:
@IEvangelist Yes the examples are of code that doesn't compile, but it's supposed to generate a **specific** compile error. In this specific one, we need CS8410, but with StreamReader it generates CS8417. I'm not sure why the compiler team made these two different error code while they look just the same for me. It's unclear to me when the compiler would generate CS8410 and when it would generate CS8417.

Reply:
Let me see what I can dig up

Reply:
Hey @Youssef1313 - try this [SharpLab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGArAbgFgAoEnANiwCYtsB2DEgbxIzdspQA4tKBZCAEsAdgAoAlI2LsMLaTPYoAnFkwA3CAgwI4EACaIMAXgzC4AdwwBJMAAcANoIDGgmACVdBhBKLy2AXxJA0mIUGhsHZ1cPfURmIKA===).

Reply:
@IEvangelist, Your example shows the error message we want here. But any idea why StreamReader generates a different message?

Reply:
The reason is that the `StreamReader` implements `IDisposable`, this goes back to my earlier comment that removing the `await` is not valid for `8410` as a suggestion to fix it. That only applies to 8417.

[See this](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQBgAIUMBWAbl1xIDZi1iMB2Q3Ab10K8bpQA5idALIQAlgDsAFAEpWOboQ7yF3FAE5iRAG4QEhBHAgATRIQC8hcXADuhAEqGTCGZWUriGgoR16joqAAOAPZQEMAANnAOxqYWVrYAIv7BoRFRjogunO5yOZreuoQAHuaWNvYZztKueV4+hACepfGESYEhYZHRTlluhAC+uIM41AzdpmwDVHgMbSmd6TF6IIQAknMdaezZXCgAzMQALK3JIXAyilM4/UA=).

```csharp
using System;
using System.Threading.Tasks;

static class Program 
{
    static async Task Main() 
    {
        await using var reader = new Reader();
        await using var disposableReader = new DisposableReader();
        
        using var x = new Reader();
        using var y = new DisposableReader();
    }
}

class Reader { }

class DisposableReader : IDisposable
{
    public void Dispose() { }
}
```